visionOS: Enable Dynamic Content Scaling for GPU process DOM rendering

hortont424 opened the pull request:
#### 597104441dc65adb03dbfe939e55869b44d685f5
<pre>
visionOS: Enable Dynamic Content Scaling for GPU process DOM rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=266058">https://bugs.webkit.org/show_bug.cgi?id=266058</a>
<a href="https://rdar.apple.com/114834211">rdar://114834211</a>

Reviewed by Said Abou-Hallawa.

Enable Dynamic Content Scaling in the GPU process, and, as this was the last
thing blocking DOM GPU process enablement on visionOS, turn it on!

The general flow is:

- the Web Content process makes a normal WebCore Display List streaming ImageBuffer,
  streaming drawing commands to the GPU process.

- the GPU process now makes a DynamicContentScalingBifurcatedImageBuffer,
  which contains the BifurcatedGraphicsContext that used to just sit directly
  in RemoteLayerBackingStore, and bifurcates the drawing commands between
  the bitmap backing store and a DynamicContentScalingImageBufferBackend
  that records the long-lived display list.

- when encoding the front buffer&apos;s contents, we synchronously grab the display
  list from the GPU process. This isn&apos;t ideal, because it means we encode them
  late, and serially, but it matches the prior design; we can make improvements
  subsequently.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Turn on DOM GPUP on visionOS.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
Allow BifurcatedGraphicsContext to be allocated on the heap; this was an aspirational
forbiddance (because BifurcatedGraphicsContext holds two GraphicsContext references),
but now we&apos;ll just have to depend on people doing the right thing (at least until
painting code gets smart-pointer-ified).

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::ImageBuffer):
Pass the ImageBufferCreationContext down to the next level of the ImageBuffer
constructor, so we can use it when making the bifurcated backend.

(WebCore::ImageBuffer::dynamicContentScalingDisplayList):
Add a failable getter for the display list, like we have for IOSurface and Cairo Surfaces.
After much deliberation, we settled on this approach to extract the display list
instead of sending backend handles back to the WP ahead of time; this turns out to be
infinitely simpler.

* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/RenderingMode.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::surfaceNameToNSString):
(WebCore::IOSurface::nameForRenderingPurpose):
Add a RenderingPurpose for &quot;bitmap only&quot; layers, as a way to indicate semi-generically
to the GPU process that the given layer doesn&apos;t need a display list.

* Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h: Renamed from Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingDisplayList.h.
(WebCore::DynamicContentScalingDisplayList::DynamicContentScalingDisplayList):
(WebCore::DynamicContentScalingDisplayList::displayList const):
(WebCore::DynamicContentScalingDisplayList::takeSurfaces):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Move DynamicContentScalingDisplayList to WebCore.
Generate its encoders.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isDynamicContentScalingEnabled const):
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
Plumb the &quot;dynamic content scaling enabled&quot; preference to the GPU process.

* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::dynamicContentScalingDisplayList):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::dynamicContentScalingDisplayList):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
Plumb the display list back to the Web Content process when requested.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::createImageBufferInternal):
(WebKit::RemoteRenderingBackend::createImageBuffer):
Factor out creation of the primary image buffer, and override the container type
if we need a bifurcated ImageBuffer (if the preference is enabled and the buffer
is for non-bitmap-only layer backing store).

* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h: Added.
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm: Added.
(WebKit::DynamicContentScalingBifurcatedImageBuffer::DynamicContentScalingBifurcatedImageBuffer):
(WebKit::DynamicContentScalingBifurcatedImageBuffer::context const):
(WebKit::DynamicContentScalingBifurcatedImageBuffer::dynamicContentScalingDisplayList):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::createBackendHandle const):
Add DynamicContentScalingBifurcatedImageBuffer, which just overrides
context() to create a BifurcatedGraphicsContext and records the display list.

Clear the context after encoding the display list; this replaces the releaseGraphicsContext()
that used to be in swapToValidFrontBuffer. This ensures that we never accumulate display lists.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
(WebKit::hasValue):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStoreProperties::layerContentsBufferFromBackendHandle):
(WebKit::RemoteLayerBackingStore::drawInContext): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::allocateBufferInternal):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::allocateBuffer const):
Remove all of the existing support for bifurcation directly in RemoteLayerBackingStore
and RemoteLayerWithInProcessRenderingBackingStore, adopting
DynamicContentScalingBifurcatedImageBuffer for both in-process and GPUP layers.
Plumb the &quot;containsBitmapOnly&quot; bit along in the RenderingPurpose.

* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.mm:
(-[WKCompositingLayer _setWKContents:withDisplayList:replayForTesting:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandleSharing.h:
(WebKit::ImageBufferBackendHandleSharing::dynamicContentScalingDisplayList):
(WebKit::ImageBufferBackendHandleSharing::setBackendHandle):
(WebKit::ImageBufferBackendHandleSharing::hasBackendHandle const): Deleted.
Minor adjustments for new names.

* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::hasBackendHandle const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
Remove hasBackendHandle, which is unused.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::containsBitmapOnly const):
(WebKit::PlatformCALayerRemote::shouldIncludeDisplayListInBackingStore const):
Expose containsBitmapOnly() for use everywhere we compute the RenderingPurpose.

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::shouldUseRemoteRenderingFor):
Remove the default case and explicitly list all values, so you can&apos;t get tripped
up by not populating this switch when you add new ones.

Canonical link: <a href="https://commits.webkit.org/271770@main">https://commits.webkit.org/271770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3a5c2da7cb6109ca2df7c0d2429dd0301bd0858

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32113 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30215 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10410 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5548 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29877 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5891 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29757 "webkitpy-tests (failure)") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33455 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27002 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26773 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7713 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36174 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7021 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7797 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
```diff
@@ -35,7 +35,6 @@ namespace WebCore {
 // Any state that is returned by GraphicsContext methods will be retrieved from the primary context.
 
 class WEBCORE_EXPORT BifurcatedGraphicsContext : public GraphicsContext {
-    WTF_FORBID_HEAP_ALLOCATION;
```
Should it be `WTF_MAKE_NONCOPYABLE`?

hortont424 replied:
Sure!

shallawa commented on Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h:
```diff
@@ -27,36 +27,27 @@
     DynamicContentScalingDisplayList() = default;
-    DynamicContentScalingDisplayList(WebCore::SharedBuffer& displayList, Vector<MachSendRight>&& surfaces)
-        : m_displayList(&displayList)
+    DynamicContentScalingDisplayList(RefPtr<WebCore::SharedBuffer> displayList, Vector<MachSendRight>&& surfaces)
```
Why this change is needed? The reference was guaranteeing `displayList` will be valid. Can this class holds a null `m_displayList`?

hortont424 replied:
This was just required to make the generated serializers work. The contract remains (I can add an assert or something? or maybe Alex knows how to fix it)

hortont424 replied:
(but I'll try ref)

hortont424 replied:
ref works

shallawa commented on Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h:
```diff
@@ -27,36 +27,27 @@
-    DynamicContentScalingDisplayList(WebCore::SharedBuffer& displayList, Vector<MachSendRight>&& surfaces)
-        : m_displayList(&displayList)
+    DynamicContentScalingDisplayList(RefPtr<WebCore::SharedBuffer> displayList, Vector<MachSendRight>&& surfaces)
+        : m_displayList(displayList)
```
Should we use WTFMove() here if we change the input to be RefPtr or Ref?

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -219,26 +224,46 @@ void RemoteRenderingBackend::moveToImageBuffer(RenderingResourceIdentifier image
-            imageBuffer = ImageBuffer::create<ImageBufferShareableMappedIOSurfaceBackend>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext, imageBufferIdentifier);
-    } else
-        imageBuffer = ImageBuffer::create<ImageBufferShareableBitmapBackend>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext, imageBufferIdentifier);
-#else
```
I think this `#else` should not be removed and the structure of this code should not be changed. See [268251@main](https://commits.webkit.org/268251@main).

hortont424 replied:
OK!

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -219,26 +224,46 @@ void RemoteRenderingBackend::moveToImageBuffer(RenderingResourceIdentifier image
+#endif
+
+    if (!imageBuffer)
+        imageBuffer = createImageBufferInternal<ImageBuffer>(logicalSize, renderingMode, purpose, resolutionScale, colorSpace, pixelFormat, creationContext, imageBufferIdentifier);
```
Is it okay if `isDynamicContentScalingEnabled()` is true but createImageBufferInternal<DynamicContentScalingBifurcatedImageBuffer> fails for any reason to return an `ImageBuffer`? Does WebProcess expects `isDynamicContentScalingEnabled()` to be respected?

Or should this code structured like this?

```
#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
    if (m_gpuConnectionToWebProcess->isDynamicContentScalingEnabled() && ...)
        imageBuffer = createImageBufferInternal<DynamicContentScalingBifurcatedImageBuffer>(...);
    else
        imageBuffer = createImageBufferInternal< ImageBuffer >(...);
#else
        imageBuffer = createImageBufferInternal< ImageBuffer >(...);
#endif
```

hortont424 replied:
AFAIK the current code is fine; if we don't end up with a DL we just don't install the DL; it's OK if isDynamicContentScalingEnabled() is not respected (in either direction!)

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h:
```diff
@@ -23,29 +23,38 @@
+#include <WebCore/DynamicContentScalingDisplayList.h>
+#include <WebCore/ImageBuffer.h>
+#include <WebCore/ImageBufferCGBackend.h>
+#include <wtf/IsoMalloc.h>
```
Do we really need to include these header files? Can't we replace some of them by forward declarations?

hortont424 replied:
Quite possibly! Will check.

hortont424 replied:
drastically reduced

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
```diff
@@ -0,0 +1,77 @@
+
+#import "DynamicContentScalingImageBufferBackend.h"
+#import <WebCore/DynamicContentScalingDisplayList.h>
+#import <WebCore/ImageBufferBackend.h>
```
Some of these headers are included in DynamicContentScalingBifurcatedImageBuffer.h.

hortont424 replied:
not anymore!

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
```diff
@@ -0,0 +1,77 @@
+
+DynamicContentScalingBifurcatedImageBuffer::DynamicContentScalingBifurcatedImageBuffer(Parameters parameters, const WebCore::ImageBufferBackend::Info& backendInfo, const WebCore::ImageBufferCreationContext& creationContext, std::unique_ptr<WebCore::ImageBufferBackend>&& backend, WebCore::RenderingResourceIdentifier renderingResourceIdentifier)
+    : ImageBuffer(parameters, backendInfo, creationContext, WTFMove(backend), renderingResourceIdentifier)
+    , m_dynamicContentScalingBackend(DynamicContentScalingImageBufferBackend::create(ImageBuffer::backendParameters(parameters), creationContext))
```
Can DynamicContentScalingImageBufferBackend::create() fail for any reason? I would suggest the following:

1. `RemoteRenderingBackend::createImageBuffer()` creates `DynamicContentScalingImageBufferBackend`. If the creation fails it will fall back to normal `ImageBuffer` or return an error.
2. `createImageBufferInternal()` will pass `dynamicContentScalingBackend` as the last argument to `ImageBuffer::create()`.
3. `ImageBuffer::create()` will pass it here to this constructor as the last argument.
4. This constructor will move it to `m_dynamicContentScalingBackend`

hortont424 replied:
Only for empty backendSize, and I'm not even sure why we do that.

I'll give this a shot.

hortont424 replied:
I don't love this, because it means passing dynamicContentScalingBackend even in the non-DCS case, or duplicating this code, or something? Going to defer this to later cleanup. Also I don't think it's critical because the main ImageBuffer backend will also fail to allocate for a superset of the reasons that the DCS one would (which is: only empty backendSize)

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
```diff
@@ -91,6 +91,8 @@ static CFDictionaryRef makeContextOptions(const DynamicContentScalingImageBuffer
 {
 }
 
+DynamicContentScalingImageBufferBackend::~DynamicContentScalingImageBufferBackend() = default;
```
Why is this needed?

hortont424 replied:
Will check if it still is; this was to avoid adding a header to the .h

hortont424 replied:
it is not

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
```diff
@@ -111,19 +112,31 @@
+    ImageBufferCreationContext creationContext;
+    creationContext.surfacePool = &WebCore::IOSurfacePool::sharedPool();
+
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes)
+        return allocateBufferInternal<DynamicContentScalingBifurcatedImageBuffer>(type(), size(), purpose, scale(), colorSpace(), pixelFormat(), creationContext);
+#endif
+
+    return allocateBufferInternal<ImageBuffer>(type(), size(), purpose, scale(), colorSpace(), pixelFormat(), creationContext);
```
Don't we need to check `isDynamicContentScalingEnabled()`? Also this code is repeated above. Can't it be shared somehow?

hortont424 replied:
`m_parameters.includeDisplayList` already checks `isDynamicContentScalingEnabled`

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
```diff
@@ -38,7 +41,7 @@ using ImageBufferBackendHandle = std::variant<
     , MachSendRight
 #endif
 #if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
-    , DynamicContentScalingDisplayList
+    , WebCore::DynamicContentScalingDisplayList
```
I think I am confused here. Why is this considered a backend handle? I do not think there is an backend backed by a DisplayList.

hortont424 replied:
Sure there is; `DynamicContentScalingImageBufferBackend`, that's existing code (and it's what `DynamicContentScalingBifurcatedImageBuffer` uses to encapsulate the DL management. You could imagine moving that code into `DynamicContentScalingBifurcatedImageBuffer` but we can do that separately.

shallawa commented on Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h:
```diff
@@ -27,38 +27,28 @@
-
 private:
-    RefPtr<WebCore::SharedBuffer> m_displayList;
+    Ref<WebCore::SharedBuffer> m_displayList;
```
Nice!

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -238,25 +243,47 @@ void RemoteRenderingBackend::moveToImageBuffer(RenderingResourceIdentifier image
+        imageBuffer = ImageBuffer::create<ImageBufferShareableBitmapBackend, ImageBufferType>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext, imageBufferIdentifier);
 #else
-    imageBuffer = ImageBuffer::create<ImageBufferShareableBitmapBackend>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext, imageBufferIdentifier);
+    if (!imageBuffer)
```
There is no need for this check. imageBuffer is for sure a nullptr.

hortont424 replied:
quite right.

hortont424 committed:
visionOS: Enable Dynamic Content Scaling for GPU process DOM rendering
https://bugs.webkit.org/show_bug.cgi?id=266058
rdar://114834211

Reviewed by Said Abou-Hallawa.

Enable Dynamic Content Scaling in the GPU process, and, as this was the last
thing blocking DOM GPU process enablement on visionOS, turn it on!

The general flow is:

- the Web Content process makes a normal WebCore Display List streaming ImageBuffer,
  streaming drawing commands to the GPU process.

- the GPU process now makes a DynamicContentScalingBifurcatedImageBuffer,
  which contains the BifurcatedGraphicsContext that used to just sit directly
  in RemoteLayerBackingStore, and bifurcates the drawing commands between
  the bitmap backing store and a DynamicContentScalingImageBufferBackend
  that records the long-lived display list.

- when encoding the front buffer's contents, we synchronously grab the display
  list from the GPU process. This isn't ideal, because it means we encode them
  late, and serially, but it matches the prior design; we can make improvements
  subsequently.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Turn on DOM GPUP on visionOS.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
Allow BifurcatedGraphicsContext to be allocated on the heap; this was an aspirational
forbiddance (because BifurcatedGraphicsContext holds two GraphicsContext references),
but now we'll just have to depend on people doing the right thing (at least until
painting code gets smart-pointer-ified).

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::ImageBuffer):
Pass the ImageBufferCreationContext down to the next level of the ImageBuffer
constructor, so we can use it when making the bifurcated backend.

(WebCore::ImageBuffer::dynamicContentScalingDisplayList):
Add a failable getter for the display list, like we have for IOSurface and Cairo Surfaces.
After much deliberation, we settled on this approach to extract the display list
instead of sending backend handles back to the WP ahead of time; this turns out to be
infinitely simpler.

* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/RenderingMode.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::surfaceNameToNSString):
(WebCore::IOSurface::nameForRenderingPurpose):
Add a RenderingPurpose for "bitmap only" layers, as a way to indicate semi-generically
to the GPU process that the given layer doesn't need a display list.

* Source/WebCore/platform/graphics/cocoa/DynamicContentScalingDisplayList.h: Renamed from Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingDisplayList.h.
(WebCore::DynamicContentScalingDisplayList::DynamicContentScalingDisplayList):
(WebCore::DynamicContentScalingDisplayList::displayList const):
(WebCore::DynamicContentScalingDisplayList::takeSurfaces):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
Move DynamicContentScalingDisplayList to WebCore.
Generate its encoders.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::isDynamicContentScalingEnabled const):
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
Plumb the "dynamic content scaling enabled" preference to the GPU process.

* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::dynamicContentScalingDisplayList):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::dynamicContentScalingDisplayList):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
Plumb the display list back to the Web Content process when requested.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::createImageBufferInternal):
(WebKit::RemoteRenderingBackend::createImageBuffer):
Factor out creation of the primary image buffer, and override the container type
if we need a bifurcated ImageBuffer (if the preference is enabled and the buffer
is for non-bitmap-only layer backing store).

* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h: Added.
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm: Added.
(WebKit::DynamicContentScalingBifurcatedImageBuffer::DynamicContentScalingBifurcatedImageBuffer):
(WebKit::DynamicContentScalingBifurcatedImageBuffer::context const):
(WebKit::DynamicContentScalingBifurcatedImageBuffer::dynamicContentScalingDisplayList):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
(WebKit::DynamicContentScalingImageBufferBackend::createBackendHandle const):
Add DynamicContentScalingBifurcatedImageBuffer, which just overrides
context() to create a BifurcatedGraphicsContext and records the display list.

Clear the context after encoding the display list; this replaces the releaseGraphicsContext()
that used to be in swapToValidFrontBuffer. This ensures that we never accumulate display lists.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
(WebKit::hasValue):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStoreProperties::layerContentsBufferFromBackendHandle):
(WebKit::RemoteLayerBackingStore::drawInContext): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::allocateBufferInternal):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::swapToValidFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::allocateBuffer const):
Remove all of the existing support for bifurcation directly in RemoteLayerBackingStore
and RemoteLayerWithInProcessRenderingBackingStore, adopting
DynamicContentScalingBifurcatedImageBuffer for both in-process and GPUP layers.
Plumb the "containsBitmapOnly" bit along in the RenderingPurpose.

* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/cocoa/RemoteLayerTreeLayers.mm:
(-[WKCompositingLayer _setWKContents:withDisplayList:replayForTesting:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandleSharing.h:
(WebKit::ImageBufferBackendHandleSharing::dynamicContentScalingDisplayList):
(WebKit::ImageBufferBackendHandleSharing::setBackendHandle):
(WebKit::ImageBufferBackendHandleSharing::hasBackendHandle const): Deleted.
Minor adjustments for new names.

* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::hasBackendHandle const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
Remove hasBackendHandle, which is unused.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::containsBitmapOnly const):
(WebKit::PlatformCALayerRemote::shouldIncludeDisplayListInBackingStore const):
Expose containsBitmapOnly() for use everywhere we compute the RenderingPurpose.

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::shouldUseRemoteRenderingFor):
Remove the default case and explicitly list all values, so you can't get tripped
up by not populating this switch when you add new ones.

Canonical link: https://commits.webkit.org/271770@main
